### ISSUE TITLE
Highlighting feature value colors when using KernelExplainer with TFIDFVectorizer

### ISSUE THREAD

**Issue by Ravikiran-Bhat-Git** [shap/shap#364]
I am trying to make a summary plot using KernelExplainer with TFIDFVectorizer as feature extractor for text classifiation.  I am trying to obtain the summary plot as follows:
`x_train, x_test, y_train, y_test = train_test_split(list_corpus, list_labels, test_size=0.3, random_state=40)`

`vec = TfidfVectorizer(sublinear_tf=True,  strip_accents='unicode', analyzer='word', smooth_idf=True,  ngram_range=(1, 3))`

`vec.fit(x_train)`
`estimator = LogisticRegression(solver='newton-cg', n_jobs=-1)`
`train_vectors = vec.fit_transform(x_train)`
`test_vectors = vec.transform(x_test)`
`estimator.fit(train_vectors, y_train)`
`explainer = shap.KernelExplainer(estimator.predict_proba,  train_vectors)`

`shap_values = explainer.shap_values(test_vectors)`
`feature_names = vec.get_feature_names()`
`shap.summary_plot(shap_values[0],feature_names=feature_names)`

Currently the output I am getting is as follows:
![image](https://user-images.githubusercontent.com/12827109/50374409-138eed00-05ee-11e9-9b13-3228a64a46f9.png)
The summary plot does not  include the feature values in color. I tried to use the matrix given by vec.fit_transform(x_train) as a parameter for the summary plot but that didn't help either. Is there any way to highlight the feature values in this plot.

**Comment by Ravikiran-Bhat-Git** [comment]
`x_train, x_test, y_train, y_test = train_test_split(list_corpus, list_labels, test_size=0.3, random_state=40)`

`vec = TfidfVectorizer(sublinear_tf=True,  strip_accents='unicode', analyzer='word', smooth_idf=True,  ngram_range=(1, 3))`

`vec.fit(x_train)`
`estimator = LogisticRegression(solver='newton-cg', n_jobs=-1)`
`train_vectors = vec.fit_transform(x_train)`
`test_vectors = vec.transform(x_test)`
`estimator.fit(train_vectors, y_train)`
`explainer = shap.KernelExplainer(estimator.predict_proba,  train_vectors)`

`shap_values = explainer.shap_values(test_vectors)`
`feature_names = vec.get_feature_names()`
`shap.summary_plot(shap_values[0],feature_names=feature_names)`

**Comment by slundberg** [comment]
I looked into it and it is because the data matrix was sparse. If you just do `.toarray()` before passing the data matrix to summary_plot it should work.

Also, since you are explaining a linear model you should consider using the LinearExplainer since it will compute the same values as KernelExplainer (if explaining the log-odds model output) but quickly and exactly. I just posted a notebook to give an example of using it (note it depends on v0.27 which is only on github at the moment not yet on pypi): https://github.com/slundberg/shap/blob/master/notebooks/linear_explainer/Sentiment%20Analysis%20with%20Logistic%20Regression.ipynb

**Comment by Ravikiran-Bhat-Git** [comment]
Thanks. I will give it a try.

**Comment by Ravikiran-Bhat-Git** [comment]
The summary plot worked with `.toarray()`.  The output I am getting for 
`shap.summary_plot(shap_values[0],test_vectors.toarray(),feature_names=vec.get_feature_names())` is as follows:
![all_features_0](https://user-images.githubusercontent.com/12827109/50387387-b765bf00-06fa-11e9-8aa7-4c26593fac84.png)
How do I interpret this output. Since I currently have 2 classes, what does a negative SHAP value for a feature (say for example the "desired software") mean from this context. When plotting the summary plot as:
`shap.summary_plot(shap_values[1],test_vectors.toarray(),feature_names=vec.get_feature_names())`
we get a flipped version of the first plot:
![all_features_1](https://user-images.githubusercontent.com/12827109/50387412-791ccf80-06fb-11e9-8b26-4656ae8c5882.png)
Does this indicate that features having a negative impact on the first class will have positive impact in the second. Is this interpretation correct?

**Comment by slundberg** [comment]
That's right, since you have two classes the log-odds of one class is perfectly determined by the log-odds of the other class. So you really have two redundant explanations, one for each class.

A small point, if I were you I would explain 1000 samples (or more) for the summary plots because then you will see much clearer patterns (and LinearExplainer is fast).

**Comment by Ravikiran-Bhat-Git** [comment]
Thanks for the explanation. One more question I had was the SHAP value of 0 for some features. Consider the following plot for **Class 0** when I use Naive Bayes classifier instead of LogisticRegression. Features such as "software" and "product" show that some samples have high value for these features which is consistent with my observation of the data. But the plot also indicates that for these samples they have 0 SHAP value (which indicates 0 impact on model output).
![all_features_0_nb](https://user-images.githubusercontent.com/12827109/50603458-d1ea2600-0eba-11e9-8238-6211e9f9d061.png)
 
Theoretically, these features (and also including several others in the plot) highlighted in red should have some positive impact towards the **Class 0**.  Am i mistaken in my interpretation of this SHAP value or is this a natural consequence resulting due to the computation of the SHAP value involving all possible coalition of the features?

On a smaller note, the reason I am using a small number of samples and KernelExplainer is because I am working on my thesis where I am constrained to work on a dataset of very small size as well as requiring model agnostic explanations that indicate global as well as locally important features and their impacts on model decisions.

**Comment by slundberg** [comment]
A note on small sample sizes, as you probably already aware, using bootstrap resampling to estimate variability is a good idea.

Having an impact of exactly zero would happen if you are using the KernelExplainer and using `shap.kmeans` to pass a single mean value for the background. This is because `shap.kmeans` nudges the mean values to match the nearest value in the dataset, and if the background value and the current value of a feature are the same then it will have 0 impact. Hope that's helpful.

**Comment by kiransp027** [comment]
@slundberg Dear Slundberg, Do the text plot also could be used with TFiDF Vectorizer? or only tokenizer works with that?
